### ISSUE TITLE
Dropped input characters and cursor placement issues around closing parentheses

### ISSUE THREAD

**Issue by sangaline** [copilot-emacs/copilot.el#228]
I've been consistently seeing some strange behavior around closing parentheses and curly brackets. Here's a gif that demonstrates it:

![copilot](https://github.com/copilot-emacs/copilot.el/assets/8236564/917deeaa-fcc4-40a0-88b8-ec50b206699c)

On line 6, I type `myFunction();` but the closing `)` is dropped and I get `myFunction(;`. I then insert a `)` between `(` and `;` but the `)` is put after the `;` instead of before it. Then on line 8 I try typing `myFunction());` to make up for the `)` being dropped when typed a single time, and it adds back the dropped `)` in addition to the second one so I end up with `))`. This creates a difficult situation because it means there's no way to close brackets other than going back and fixing things after they've been entered incorrectly.

I'm using a relatively vanilla Doom Emacs install with electric-pair-mode and snippets disabled. I'm on the latest [1542d76](https://github.com/copilot-emacs/copilot.el/commit/1542d76909636bf8804dd9c770f28818a897cfdc) commit of copilot-emacs. Things behave normally if I disable copilot-mode, but I'm not sure if it's an interaction with another package or an issue with copilot-mode itself. Happy to provide any further debugging info if you let me know what would be relevant.

**Comment by emil-vdw** [comment]
This must be from the [jumping cursor bugfix](https://github.com/copilot-emacs/copilot.el/commit/1542d76909636bf8804dd9c770f28818a897cfdc). I think it doesn't play well with `electric-pair-mode`.

**Comment by matthemsteger** [comment]
This latest change made this library unusable also if you use https://github.com/Fuco1/smartparens. I don't know the usage percentage of electric-pair-mode and smartparens, but the jumping cursor bug was a lot less of an issue at least for me (in fact I dont think I noticed it before).

With smartparens, if you type a character that is part of the copilot complete, it closes the paren (or curly or whatever) and continues completing and you end up with a garbled mess.

Going back to the commit before the above jumping cursor fix is a lot smoother experience and make copilot usable in emacs + smartparens again.

**Comment by emil-vdw** [comment]
@sangaline, @matthemsteger it must indeed be very annoying. This bug will have the highest priority for me and I hope to have it fixed before end of next week.

**Comment by jimeh** [comment]
I'm having the same issue after #222 with `smartparens-mode`. In case it helps, here's a demo:

https://github.com/copilot-emacs/copilot.el/assets/39563/0be8b66f-9a3e-4527-bf1c-89688425af46

**Comment by martin-dimi** [comment]
Same here. This is a very annoying bug. Had to stop using copilot altogether because of it. Please prioritise 🙏

**Comment by emil-vdw** [comment]
@martin-dimi for now you can use an older commit until it is fixed.

**Comment by jimeh** [comment]
@emil-vdw I could be wrong, but this new issue caused by #222 seems worse than the issue it fixed. What do you think about reverting #222 until a fix addressing both is ready?

**Comment by emil-vdw** [comment]
@jimeh that's probably best. I will revert shortly.

**Comment by sangaline** [comment]
@emil-vdw I'm unfortunately still seeing the same behavior after the revert. Unlike some others here, I am not using electric-pair-mode or smartparens-mode. I don't have anything configured to insert closing brackets automatically.

**Comment by kliph-gladly** [comment]
I am also still seeing this issue on the latest code.  I got around it by adding 

```emacs-lisp
(add-hook 'post-self-insert-hook #'copilot-clear-overlay)
```

To update the overlay (which unfortunately flashes more than I'd like) when Electric Pair mode fires after inserting characters.
